service/collector/endpoint.go

```diff
@@ -47,23 +61,215 @@ func NewEndpoint(config EndpointConfig) (*Endpoint, error) {
+						ingressCheckSucesfull,
+						kvmConfig.Name,
+						ip,
+						ingresSchemeHttp,
+						"true",
+					)
+				}
+				if err != nil {
+					// ingress endpoint check failed
+					// ingress endpoint check was successful, proxy-protocol enabled
+					ch <- prometheus.MustNewConstMetric(
+						endpointLabelsDesc,
+						prometheus.GaugeValue,
+						ingressCheckFailure,
+						kvmConfig.Name,
+						ip,
+						ingresSchemeHttp,
+						"unknown",
+					)
+				}
+			}
+		}
+	}
 
 	return nil
 }
 
-func (i *Endpoint) Describe(ch chan<- *prometheus.Desc) error {
+func (e *Endpoint) Describe(ch chan<- *prometheus.Desc) error {
 	ch <- endpointLabelsDesc
 	return nil
 }
+
+func getEndpointIps(endpoint *corev1.Endpoints) []string {
+	var ipList []string
+	for _, subset := range endpoint.Subsets {
+		for _, address := range subset.Addresses {
+			ipList = append(ipList, address.IP)
+		}
+	}
+
+	return ipList
+}
+
+func (e *Endpoint) buildHttpRequest(ipAddress string, scheme string, port int) (*http.Request, error) {
+	u := url.URL{
+		Host:   fmt.Sprintf("%s:%d", ipAddress, port),
+		Path:   "healthz",
+		Scheme: scheme,
+	}
+
+	// be sure to close idle connection after health check is finished
+	defer e.httpTransport.CloseIdleConnections()
+
+	req, err := http.NewRequest("GET", u.String(), nil)
+	if err != nil {
+		return nil, microerror.Maskf(err, "unable to construct health check request")
+	}
+
+	// close connection after health check request (the TCP connection gets
+	// closed by deferred s.tr.CloseIdleConnections()).
+	req.Header.Add("Connection", "close")
+
+	return req, nil
+}
+
+// ingressEndpointUpClassicHttp send http packet to the endpoint to ensure target ingress endpoint ip is up
+func (e *Endpoint) ingressEndpointUpClassicHttp(ipAddress string, scheme string, port int) error {
+	req, err := e.buildHttpRequest(ipAddress, scheme, port)
+	if err != nil {
+		// failed to build http req
+		return microerror.Mask(err)
+	}
+
+	// send request to http endpoint
+	_, err = e.httpClient.Do(req)
+	if err != nil {
+		// ingress endpoint failed to respond properly
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+// ingressEndpointUpProxyProtocol send encapsulated http packet into proxy-protocol to the endpoint to ensure target ingress endpoint ip is up
+func (e *Endpoint) ingressEndpointUpProxyProtocol(ipAddress string, scheme string, port int) error {
+	var buffer bytes.Buffer
+	// build http request
+	req, err := e.buildHttpRequest(ipAddress, scheme, port)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	// write proxy-protocol header to buffer
+	_, err = fmt.Fprintf(&buffer, "PROXY TCP4 %s %s 80 80\r\n", e.localIP, e.localIP)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	// write http request to buffer
+	err = req.Write(&buffer)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// open tcp connection to ingress endpoint
+	conn, err := net.Dial("tcp", fmt.Sprintf("%s:%d", ipAddress, port))
```

Comment:
Is there any chance for endpoints to hang? We should have some timeout behaviour to ensure the whole collector is not screwed over by single endpoint failures. It would then make sense to also collect metrics for these timeouts.

Reply:
good point

Reply:
Not sure about having extra metric for timeout, if the ingress does not respond, it should be considered as fail and it dos not matter what kind of error caused it (having extra metrics for each possible error seems like a waste